src/libPMacc/include/eventSystem/tasks/TaskSetValue.hpp

```diff
@@ -82,20 +85,68 @@ getValue(T_Type& value)
 
 }
 
+/** set a value to all elements of a box
+ *
+ * @tparam T_numWorkers number of workers
+ * @tparam T_xChunkSize number of elements in x direction to prepare with one cuda block
+ */
+template<
+    uint32_t T_numWorkers,
+    uint32_t T_xChunkSize
+>
 struct KernelSetValue
 {
-    template <class DataBox, typename T_ValueType, typename Space>
-    DINLINE void operator()(DataBox data, const T_ValueType value, const Space size) const
+    /** set value to all elements
+     *
+     * @tparam T_DataBox PMacc::DataBox, type of the memory box
+     * @tparam T_ValueType type of the value
+     * @tparam T_SizeVecType PMacc::math::Vector, index type
+     *
+     * @param memBox box where all elements should be set to value
+     * @param value value to set to all elements of memBox
+     * @param size extents of memBox
+     */
+    template<
+        typename T_DataBox,
+        typename T_ValueType,
+        typename T_SizeVecType
+    >
+    DINLINE void
+    operator()(
+        T_DataBox & memBox,
+        T_ValueType const & value,
+        T_SizeVecType const & size
+    ) const
     {
-        const Space threadIndex(threadIdx);
-        const Space blockIndex(blockIdx);
-        const Space gridSize(blockDim);
-
-        Space idx(gridSize * blockIndex + threadIndex);
-
-        if (idx.x() >= size.x())
-            return;
-        data(idx) = taskSetValueHelper::getValue(value);
+        using namespace mappings::threads;
+        using SizeVecType = T_SizeVecType;
+
+        SizeVecType const blockIndex( blockIdx );
+        SizeVecType blockSize( SizeVecType::create( 1 ) );
+        blockSize.x( ) = T_xChunkSize;
+
+        constexpr uint32_t numWorkers = T_numWorkers;
+        uint32_t const workerIdx = threadIdx.x;
+
+        ForEachIdx<
+            IdxConfig<
+                T_xChunkSize,
+                numWorkers
+            >
+        >{ workerIdx }(
+            [&](
+                uint32_t const linearIdx,
+                uint32_t const
+            )
+            {
+                auto virtualWorkerIdx( SizeVecType::create( 0 ) );
+                virtualWorkerIdx.x( ) = linearIdx;
+
+                SizeVecType const idx( blockSize * blockIndex + virtualWorkerIdx );
+                if( idx.x() < size.x() )
+                    memBox( idx ) = taskSetValueHelper::getValue( value );
+            }
```

Comment:
We miss iterators in PMacc therefore we need to transform our indices in the user code.